lib/ukstore/include/uk/store.h

```diff
@@ -0,0 +1,452 @@
+	} while (0)
+
+/**
+ * Initializes a uk_store_folder_entry
+ *
+ * @param folder_entry the folder_entry to init
+ * @param new_entry the new entry to save
+ */
+#define uk_store_folder_entry_init(folder_entry, new_entry)		\
+	do {								\
+		(folder_entry)->entry = new_entry;			\
+		ukarch_store_n(&(folder_entry)->refcount.counter, 1);	\
+	} while (0)
+
+/**
+ * Initializes a uk_store_folder
+ *
+ * @param folder the folder to init
+ * @param f_name the name of the folder
+ */
+#define uk_store_folder_init(folder, f_name)				\
+	do {								\
+		(folder)->folder_name = (f_name);			\
+		UK_INIT_LIST_HEAD(&(folder)->folder_head);		\
+		UK_INIT_LIST_HEAD(&(folder)->entry_head);		\
+	} while (0)
+
+/**
+ * Adds a dynamic folder to a library
+ *
+ * @param folder the folder to add it to
+ */
+#define uk_store_add_folder(folder) \
+	_uk_store_add_folder(uk_store_libid_self(), folder)
+
+extern int
+_uk_store_add_folder(unsigned int library_id, struct uk_store_folder *folder);
+
+/**
+ * Removes a folder from a folder list
+ *
+ * @param folder pointer to the folder to remove
+ */
+#define uk_store_del_folder(folder)		\
+	uk_list_del(&(folder)->folder_head)	\
+
+/**
+ * Get a dynamic folder from a library
+ *
+ * @param folder_name the name of the folder to search for
+ * @param library_id the id returned by uk_store_libid
+ * @return the folder searched for
+ */
+#define uk_store_get_dynamic_folder(library_id, folder_name) \
+	_uk_store_get_dynamic_folder(library_id, folder_name)
+
+/**
+ * Get a static folder from a library
+ *
+ * @param folder_name the name of the folder to search for
+ * @param library_id the id returned by uk_store_libid
+ * @return the folder searched for
+ */
+#define uk_store_get_static_folder(library_id, folder_name) \
+	_uk_store_get_static_folder(library_id, folder_name)
+
+extern struct uk_store_folder *
+_uk_store_get_dynamic_folder(unsigned int library_id, const char *name);
+
+extern struct uk_store_folder *
+_uk_store_get_static_folder(unsigned int library_id, const char *name);
+
+/**
+ * Adds a folder entry to a folder list
+ *
+ * @param folder pointer to the place where to add the new entry
+ * @param folder_entry pointer to the entry to add
+ */
+#define uk_store_add_folder_entry(folder, folder_entry)	\
+	uk_list_add(&(folder_entry)->list_head, &(folder)->entry_head)
+
+/**
+ * Removes a folder entry from a folder
+ *
+ * @param folder_entry pointer to the folder entry to delete
+ */
+#define uk_store_del_folder_entry(folder_entry)	\
+	uk_list_del(&(folder_entry)->list_head)
+
+extern struct uk_store_entry *
+uk_store_get_entry(struct uk_store_folder *folder, const char *name);
+
+/**
+ * Releases an entry (decreases the refcount and sets the reference to NULL)
+ * If the refcount is 0 (it was also released by the creator), it is removed
+ * from the list and memory can be free.
+ *
+ * @param entry pointer to the entry to release
+ */
+static inline void
+uk_store_release_entry(struct uk_store_entry **entry) {
```

Comment:
If we do memory management within the library, I would then call it `free` instead of `release`